### PR TITLE
feat(filter): added filter common component

### PR DESCRIPTION
**Pull request by Hasteerp**
### Contributor checklist

- [x] This pull request is on a [separate branch](https://docs.github.com/en/get-started/quickstart/github-flow) and not the main branch
- [x] I have run the tests for the backend and frontend depending on what's needed for my changes as described in the [testing section of the contributing guide](CONTRIBUTING.md#testing)

---

### Description

I added a common reusable Filter component. 
ScreenRecording2024-11-30at12 56 16PM-ezgif com-video-to-gif-converter

The Filter component is a reusable component. 
It takes the following props: 
- HandleFilterChange
- tabs (id, name)[]
- sections (title, tags[])
     - tags: (id: number, name: string, selected: boolean)
     
This way you have full control over the tag title and tag buttons you want to populate. Additionally, decision can be made to check tab values and name based on each page requirement. It also takes handleFilterChange from the parent component that allows different filteration logic based on parent component requirement.

### Related issue

- [#550 ](https://github.com/activist-org/activist/issues/550)

### CODE REVIEW

**Comment by Hasteerp** [comment]
I'll be removing the filter component usage from organizations before merging. It is currently there for reviewer to visualize how the filter component looks.

**Comment by andrewtavis** [comment]
Looking forward to the review here, @Hasteerp! I'll try to get to this tomorrow or early next week 😊

**Comment by Hasteerp** [comment]
@cquinn540 I was having a difficult time getting the test to run. Would you have any suggestions?

**Comment by cquinn540** [comment]
> @cquinn540 I was having a difficult time getting the test to run. Would you have any suggestions?

I would need to know more about what issue you are facing.  I'll be a bit busy today, but I should have more free time to help over the next few days.

**Comment by Hasteerp** [comment]
@cquinn540 I was facing some issues trying to get my component to render I think the custom render function needs to change but I am not exactly sure. This is the error message that I got. Please feel free to check it out thanks!

`This error originated in "test/components/filter/pageFilter.spec.ts" test file. It doesn't mean the error was thrown inside the file itself, but while it was running.
The latest test that might've caused the error is "test/components/filter/pageFilter.spec.ts". It might mean one of the following:
- The error was thrown, while Vitest was running this test.
- If the error occurred after the test had been completed, this was the last documented test before it was thrown.
`

**Comment by cquinn540** [comment]
@Hasteerp Would you have time today or sooon to get on a call in Matrix chat so I can see what you are seeing?  That's the top level error message which is generic and doesn't give much information.  There's usually another stack trace somewhere in the output with more details.

**Comment by Hasteerp** [comment]
I can get on a call tomorrow if that works for you.

**Comment by cquinn540** [comment]
Sorry @Hasteerp I got into the zone working on another issue and didn't see this until now.  Let me know on the Matrix chat when you are available today.

**Comment by Hasteerp** [comment]
@andrewtavis @cquinn540 Please check it out whenever you can. I added the test case.

File: frontend/test/components/filter/pageFilter.spec.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
+import { describe, it, expect } from "vitest";
+import { render, screen } from "@testing-library/vue";
+import PageFilter from "@/components/page/filter/PageFilter.vue";
+
```
Comment (cquinn540) [inline frontend/test/components/filter/pageFilter.spec.ts:4]: Okay, I need to go back on what I said originally a little.  There is an appropriate case for a mock function here.  You should create a mock function like:
```typescript
const mockHandleFilterChange = vi.fn();
```
and you'll want to reset it after each test
```typescript
afterEach(() => {
  mockHandleFilterChange.mockClear();
});
```

File: frontend/test/components/filter/pageFilter.spec.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+          { id: 2, name: "Tag 2", selected: false },
+          { id: 3, name: "Tag 3", selected: false },
+          { id: 4, name: "Tag 4", selected: false },
+        ],
+      },
+    ],
+    tabs: [
+      { id: "tab1", name: "Tab 1" },
+      { id: "tab2", name: "Tab 2" },
+      { id: "tab3", name: "Tab 3" },
```
Comment (cquinn540) [inline frontend/test/components/filter/pageFilter.spec.ts:21]: Then you should use `mockHandleFilterChange` as the `@filter-change` prop

File: frontend/test/components/filter/pageFilter.spec.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    ],
+  };
+
+  it("should render", () => {
```
Comment (cquinn540) [inline frontend/test/components/filter/pageFilter.spec.ts:25]: Then you want a test where you click the different options and confirm that `mockHandleFilterChange` was called with the right arguments in the right order.

And another test where you tab through the options and hit enter and again confirm the mock was called correctly.

You can see the call list with [calls array](https://vitest.dev/api/mock#mock-calls) on your mock function.

Ideally you would also have a test checking that the focusing works correctly, but given your time limit I can be flexible on that.

File: frontend/test/components/filter/pageFilter.spec.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+          BtnTag: true,
+          TooltipBase: true,
+        },
+      },
+    });
+
+    expect(screen.getByText("Popular Tags")).toBeTruthy();
+    expect(screen.getByRole("heading", { name: "Filter" })).toBeTruthy();
+    expect(screen.getByLabelText("Search to filter")).toBeTruthy();
+    expect(screen.getByRole("textbox", { id: "input-search" })).toBeTruthy();
```
Comment (cquinn540) [inline frontend/test/components/filter/pageFilter.spec.ts:44]: With the other behavior tests testing for truthyness shouldn't be needed anymore

**Review by cquinn540 — CHANGES_REQUESTED** [review]

**Comment by andrewtavis** [comment]
I've made #1051 to do the follow up work to finalize the tests. We are on a bit of a deadline here for @Hasteerp, which I hope that we haven't missed, and the contributions till now are enough that we should bring them in and then continue when we have a bit more time.

File: frontend/test/components/filter/pageFilter.spec.ts (hunk 4 above)
Comment (andrewtavis) [reply frontend/test/components/filter/pageFilter.spec.ts]: Removed in my commit ✅

**Comment by Hasteerp** [comment]
@andrewtavis thanks for the pull request and allowing me to merge this one. We are currently seeing an error in linting due to unused vars (this is coming from index.vue on organizations page). This is because we commented the code out but the parameters were defined. We can either remove them completely or comment out the code for now. Please let me know what you would suggest I do and I can make a commit accordingly. Thank you for taking into consideration my deadline. Appreciate it! I'll work on the test cases in the new PR in the coming week.

**Review by andrewtavis — APPROVED** [review]
Approving these with the assumption that the most recent commit fixes all the frontend linting issues, but I'll get to them if there are any other changes.

Really appreciate this as a first contribution, @Hasteerp, and hope that we can work with you again :) Let's definitely keep #1051 in mind in the coming weeks, but this is a great start on a relatively complex component 🚀

@cquinn540: Don't mean to disregard your review, and really appreciate the support you're providing to @Hasteerp! This component isn't MVP, so under the circumstances I'd like to bring it in and iterate 😊

**Comment by cquinn540** [comment]
No problem this seems like a reasonable compromise

**Pull request merged**